lib/kontena/actors/haproxy_spawner.rb

```diff
@@ -0,0 +1,56 @@
+
+module Kontena::Actors
+  class HaproxySpawner < Concurrent::Actor::RestartingContext
+    include Kontena::Logging
+
+    ##
+    # @param [String] config_file
+    def initialize(haproxy_bin = '/usr/local/sbin/haproxy', config_file = '/etc/haproxy/haproxy.cfg')
+      @current_pid = nil
+      @haproxy_cmd = [haproxy_bin, '-f', config_file, '-db']
+      @validate_cmd = [haproxy_bin, '-c -f', config_file]
+    end
+
+    def on_message(msg)
+      case msg.action
+      when :update
+        update_haproxy
+      else
+        pass
+      end
+    end
+
+    def update_haproxy
+      if validate_config
+        if current_pid
+          reload_haproxy
+        else
+          start_haproxy
+        end
+      end
+    end
+
+    def start_haproxy
+      info 'Starting HAProxy process'
+      @current_pid = Process.spawn(@haproxy_cmd.join(' '))
+    end
+
+    def validate_config
+      system(@validate_cmd.join(' ')) == true
+    end
+
+    def reload_haproxy
+      info 'Reloading HAProxy'
+      reload_cmd = @haproxy_cmd + ['-sf', @current_pid.to_s]
+      pid = Process.spawn(reload_cmd.join(' '))
+      Process.wait(@current_pid)
```

Comment:
Why do we wait for the previous haproxy process to end? Since hproxy is "reloaded" with `-sf` flag, it could take really long for the process to actually exit as it waits till all the connections are terminated.